applications/maplab-node/src/synchronizer.cc

```diff
@@ -323,34 +333,92 @@ void Synchronizer::processOdometryMeasurement(
+          << "This might be okay during initialization.";
+
+      image_skip_counter_ += dropped_images;
+      Aligned<std::vector, vio::ImageMeasurement::ConstPtr> extracted_images;
+      image_buffer.extractItemsBeforeIncluding(
+          newest_timestamp_ns, &extracted_images);
+      all_extracted_images.insert(
+          all_extracted_images.end(), extracted_images.begin(),
+          extracted_images.end());
+    }
+  }
 
-    // Drop these frames, since there is no odometry data anymore for them.
-    const size_t dropped_nframes =
-        nframe_buffer_.removeItemsBefore(oldest_timestamp_ns);
-    LOG_IF(WARNING, dropped_nframes != 0u)
-        << "[MaplabNode-Synchronizer] Could not find an odometry "
-        << "transformation for " << dropped_nframes << " nframes "
-        << "because it was already dropped from the buffer! "
-        << "This might be okay during initialization.";
+  CHECK(visual_pipeline_) << "[MaplabNode-Synchronizer] The visual pipeline, "
+                          << "which turns individual images "
+                          << "into NFrames, has not been initialized yet!";
+
+  VLOG(5) << "[MaplabNode-Synchronizer] releasing "
+          << all_extracted_images.size()
+          << " camera images into the visual processing pipeline for "
+          << "synchronization and tracking.";
+
+  // Insert all images that are released into the visual pipeline to synchronize
+  // and track features.
+  for (const vio::ImageMeasurement::ConstPtr& image_measurement :
+       all_extracted_images) {
+    CHECK_LE(image_measurement->timestamp, newest_timestamp_ns);
+    CHECK_GE(image_measurement->timestamp, oldest_timestamp_ns);
+
+    VLOG(5)
+        << "[MaplabNode-Synchronizer] release camera image at "
+        << aslam::time::timeNanosecondsToString(image_measurement->timestamp)
+        << " into visual processing pipeline... (blocking if pipeline is full)";
+    if (!visual_pipeline_->processImageBlockingIfFull(
+            image_measurement->camera_index, image_measurement->image,
+            image_measurement->timestamp,
+            FLAGS_vio_nframe_sync_max_queue_size)) {
+      LOG(ERROR)
+          << "[MaplabNode-Synchronizer] Failed to process an image of camera "
+          << image_measurement->camera_index << " into an NFrame at time "
+          << image_measurement->timestamp << "ns!";
+      shutdown();
+    }
+    VLOG(5) << "[MaplabNode-Synchronizer] done.";
+  }
+
+  // Release all visual frames that are ready.
+  Aligned<std::vector, aslam::VisualNFrame::Ptr> finished_nframes;
+  aslam::VisualNFrame::Ptr next_nframe = visual_pipeline_->getNext();
+  size_t num_nframes_released = 0u;
+  while (next_nframe) {
+    const int64_t current_frame_timestamp_ns =
+        next_nframe->getMinTimestampNanoseconds();
 
-    frame_skip_counter_ += dropped_nframes;
+    VLOG(5) << "[MaplabNode-Synchronizer] retrieved finished VisualNFrame at "
```

Comment:
minor: if verbosity is < 5, `current_frame_timestamp_ns` is not used in this loop